cirq-core/cirq/ops/boolean_hamiltonian.py

```diff
@@ -0,0 +1,337 @@
+                hamiltonian = hamiltonian * sub_hamiltonian
+        elif isinstance(boolean_expr, Not):
+            assert len(sub_hamiltonians) == 1
+            hamiltonian = _Hamiltonian_I() - sub_hamiltonians[0]
+        elif isinstance(boolean_expr, Or):
+            hamiltonian = _Hamiltonian_O()
+            for sub_hamiltonian in sub_hamiltonians:
+                hamiltonian = hamiltonian + sub_hamiltonian - hamiltonian * sub_hamiltonian
+        elif isinstance(boolean_expr, Xor):
+            hamiltonian = _Hamiltonian_O()
+            for sub_hamiltonian in sub_hamiltonians:
+                hamiltonian = hamiltonian + sub_hamiltonian - 2.0 * hamiltonian * sub_hamiltonian
+        return hamiltonian
+
+    raise ValueError(f'Unsupported type: {type(boolean_expr)}')
+
+
+def _gray_code_comparator(k1: Tuple[int, ...], k2: Tuple[int, ...], flip: bool = False) -> int:
+    """Compares two Gray-encoded binary numbers.
+
+    Args:
+        k1: A tuple of ints, representing the bits that are one. For example, 6 would be (1, 2).
+        k2: The second number, represented similarly as k1.
+        flip: Whether to flip the comparison.
+
+    Returns:
+        -1 if k1 < k2 (or +1 if flip is true)
+        0 if k1 == k2
+        +1 if k1 > k2 (or -1 if flip is true)
+    """
+    max_1 = k1[-1] if k1 else -1
+    max_2 = k2[-1] if k2 else -1
+    if max_1 != max_2:
+        return -1 if (max_1 < max_2) ^ flip else 1
+    if max_1 == -1:
+        return 0
+    return _gray_code_comparator(k1[0:-1], k2[0:-1], not flip)
+
+
+def _get_gates_from_hamiltonians(
+    hamiltonian_polynomial_list: List['cirq.PauliSum'],
+    qubit_map: Dict[str, 'cirq.Qid'],
+    theta: float,
+    ladder_target: bool = False,
+):
+    """Builds a circuit according to [1].
+
+    Args:
+        hamiltonian_polynomial_list: the list of Hamiltonians, typically built by calling
+            _build_hamiltonian_from_boolean().
+        qubit_map: map of string (boolean variable name) to qubit.
+        theta: A single float scaling the rotations.
+        ladder_target: Whether to use convention of figure 7a or 7b.
+
+    Yield:
+        Gates that are the decomposition of the Hamiltonian.
+    """
+    combined: 'cirq.PauliSum' = sum(hamiltonian_polynomial_list, _Hamiltonian_O())
+
+    qubit_names = sorted(qubit_map.keys())
+    qubits = [qubit_map[name] for name in qubit_names]
+    qubit_indices = {qubit: i for i, qubit in enumerate(qubits)}
+
+    hamiltonians = {}
+    for pauli_string in combined:
+        w = pauli_string.coefficient.real
+        qubit_idx = tuple(sorted(qubit_indices[qubit] for qubit in pauli_string.qubits))
+        hamiltonians[qubit_idx] = w
+
+    # Here we follow improvements of [4] cancelling out the CNOTs. The first step is to order by
+    # Gray code so that as few as possible gates are changed.
+    sorted_hs = sorted(list(hamiltonians.keys()), key=functools.cmp_to_key(_gray_code_comparator))
+
+    def _simplify_cnots(cnots):
```

Comment:
This method should be pulled out, there is no good reason to be nested. 
Also, it should be simplified/refactored, the three layers of nested loops is hard to understand.

Reply:
Split in three functions, with comments.